cmd/api.go

```diff
@@ -18,157 +18,205 @@ package main
+
 	if os.Getenv("KUBEBUILDER_ENABLE_PLUGINS") != "" {
 		cmd.Flags().StringVar(&o.pattern, "pattern", "",
 			"generates an API following an extension pattern (addon)")
 	}
-	cmd.Flags().BoolVar(&o.apiScaffolder.Force, "force", false,
+
+	cmd.Flags().BoolVar(&o.force, "force", false,
 		"attempt to create resource even if it already exists")
-	o.apiScaffolder.Resource = resourceForFlags(cmd.Flags())
-}
 
-// resourceForFlags registers flags for Resource fields and returns the Resource
-func resourceForFlags(f *flag.FlagSet) *resource.Resource {
-	r := &resource.Resource{}
-	f.StringVar(&r.Kind, "kind", "", "resource Kind")
-	f.StringVar(&r.Group, "group", "", "resource Group")
-	f.StringVar(&r.Version, "version", "", "resource Version")
-	f.BoolVar(&r.Namespaced, "namespaced", true, "resource is namespaced")
-	f.BoolVar(&r.CreateExampleReconcileBody, "example", true,
+	o.resource = &resource.Resource{}
+	cmd.Flags().StringVar(&o.resource.Kind, "kind", "", "resource Kind")
+	cmd.Flags().StringVar(&o.resource.Group, "group", "", "resource Group")
+	cmd.Flags().StringVar(&o.resource.Version, "version", "", "resource Version")
+	cmd.Flags().BoolVar(&o.resource.Namespaced, "namespaced", true, "resource is namespaced")
+	cmd.Flags().BoolVar(&o.resource.CreateExampleReconcileBody, "example", true,
 		"if true an example reconcile body should be written while scaffolding a resource.")
-	return r
 }
 
-// APICmd represents the resource command
-func (o *apiOptions) runAddAPI() {
-	internal.DieIfNotConfigured()
-
-	switch strings.ToLower(o.pattern) {
-	case "":
-		// Default pattern
-
-	case "addon":
-		o.apiScaffolder.Plugins = append(o.apiScaffolder.Plugins, &addon.Plugin{})
-
-	default:
-		log.Fatalf("unknown pattern %q", o.pattern)
+func (o *apiOptions) loadConfig() (*config.Config, error) {
+	projectConfig, err := config.Load()
+	if os.IsNotExist(err) {
+		return nil, errors.New("unable to find configuration file, project must be initialized")
 	}
 
-	if err := o.apiScaffolder.Validate(); err != nil {
-		log.Fatalln(err)
+	return projectConfig, err
+}
+
+func (o *apiOptions) validate(c *config.Config) error {
+	if err := o.resource.Validate(); err != nil {
+		return err
 	}
 
 	reader := bufio.NewReader(os.Stdin)
 	if !o.resourceFlag.Changed {
 		fmt.Println("Create Resource [y/n]")
-		o.apiScaffolder.DoResource = util.YesNo(reader)
+		o.doResource = internal.YesNo(reader)
 	}
-
 	if !o.controllerFlag.Changed {
 		fmt.Println("Create Controller [y/n]")
-		o.apiScaffolder.DoController = util.YesNo(reader)
-	}
-
-	fmt.Println("Writing scaffold for you to edit...")
-
-	if err := o.apiScaffolder.Scaffold(); err != nil {
-		log.Fatal(err)
+		o.doController = internal.YesNo(reader)
 	}
 
-	if err := o.postScaffold(); err != nil {
-		log.Fatal(err)
-	}
-}
+	// In case we want to scaffold a resource API we need to do some checks
+	if o.doResource {
+		// Skip the following check for v1 as resources aren't tracked
+		if !c.IsV1() {
```

Comment:
nit: maybe this is better in terms of readability.
```
		if !c.IsV1() {
                	return nil
                }
		// Check that resource doesn't exist or flag force was set
		if !o.force {
			resourceExists := false
			for _, r := range c.Resources {
				if r.Group == o.resource.Group &&
					r.Version == o.resource.Version &&
					r.Kind == o.resource.Kind {
					resourceExists = true
					break
				}
			}
			if resourceExists {
				return errors.New("API resource already exists")
			}
		}
```

Reply:
I wanted to make the difference from things that were `v2`-specific, like `multigroup` from things that are just not in `v1`. In case of a potential `v3`, `multigroup` may be supported straight away, but I'm quite sure that resources will net stop being tracked. Does that make sense?

Reply:
Fair enough